File: smallpict-php/src/Laravel/Facades/SmallPict.php
<?php

declare(strict_types=1);

namespace SmallPict\Laravel\Facades;

use SmallPict\Client;

/**
 * @method static \SmallPict\Models\OptimizeResult optimize(string|resource $source, ?\SmallPict\Models\OptimizeOptions $options = null)
 * @method static \SmallPict\Models\QuotaResponse getQuota()
 * @method static \SmallPict\Models\PurgeResponse purgeCdn(array|string $urls = [], string $purgeType = 'url')
 * @method static bool validateKey()
 * @method static \SmallPict\Models\JobStatusResult getJobStatus(string $jobId)
 *
 * @see \SmallPict\Client
 */
class SmallPict
{
    /**
     * @param string $method
     * @param array<mixed> $args
     * @return mixed
     */
    public static function __callStatic(string $method, array $args)
    {
        if (function_exists('app')) {
            $instance = app(Client::class);
            return $instance->$method(...$args);
        }
        throw new \RuntimeException('Laravel container is not available');
    }
}

File: smallpict-ruby/lib/smallpict.rb
# frozen_string_literal: true

require_relative "smallpict/version"
require_relative "smallpict/configuration"
require_relative "smallpict/crypto"
require_relative "smallpict/errors"
require_relative "smallpict/models/optimize_options"
require_relative "smallpict/models/optimize_result"
require_relative "smallpict/models/quota_response"
require_relative "smallpict/models/purge_response"
require_relative "smallpict/models/job_status_result"
require_relative "smallpict/client"
require_relative "smallpict/active_storage/service"

module SmallPict
  class << self
    def configuration
      @configuration ||= Configuration.new
    end

    def configure
      yield(configuration)
    end

    def reset_configuration!
      @configuration = Configuration.new
      @default_client = nil
    end

    def client
      @default_client ||= Client.new
    end

    def optimize(source, options = {})
      client.optimize(source, options)
    end

    def get_quota
      client.get_quota
    end

    def purge_cdn(urls = [], purge_type: "url")
      client.purge_cdn(urls, purge_type: purge_type)
    end

    def validate_key
      client.validate_key
    end

    def get_job_status(job_id)
      client.get_job_status(job_id)
    end
  end
end
